### ISSUE TITLE
Light-speed pig motion when loading some multiverse worlds

### ISSUE THREAD

**Issue by PikaMug** [CitizensDev/Citizens2#2810]
### `/version` output

```txt
This server is running Paper version git-Paper-"1790528" (MC: 1.18.2) (Implementing API version 1.18.2-R0.1-SNAPSHOT) (Git: 1790528)
```

### `/version Citizens` output

```txt
Citizens version 2.0.30-SNAPSHOT (build 2565)
```

### Log file

https://paste.denizenscript.com/View/97400

### config.yml

https://paste.denizenscript.com/View/97401

### saves.yml

https://paste.denizenscript.com/View/97402

### Description

Server runs out of memory (6GB, typically utilizes 4) from a Pig with waypoints. The error is consistent across restarts.

```
[01:30:16] [Server thread/ERROR]: Entity threw exception at hub:1113.161624536367,77.0,-717.4499999880791
java.lang.OutOfMemoryError: Java heap space
    at org.bukkit.craftbukkit.v1_18_R2.util.UnsafeList.growIfNeeded(UnsafeList.java:231) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at org.bukkit.craftbukkit.v1_18_R2.util.UnsafeList.add(UnsafeList.java:68) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at io.papermc.paper.util.CollisionUtil.getCollisionsForBlocksOrWorldBorder(CollisionUtil.java:491) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at io.papermc.paper.util.CollisionUtil.getCollisions(CollisionUtil.java:857) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at net.minecraft.world.entity.Entity.collide(Entity.java:1307) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at net.minecraft.world.entity.Entity.move(Entity.java:976) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at net.minecraft.world.entity.LivingEntity.handleRelativeFrictionAndCalculateMovement(LivingEntity.java:2814) ~[?:?]
    at net.minecraft.world.entity.LivingEntity.travel(LivingEntity.java:2769) ~[?:?]
    at net.minecraft.world.entity.animal.Pig.travelWithInput(Pig.java:282) ~[?:?]
    at net.minecraft.world.entity.ItemSteerable.travel(ItemSteerable.java:25) ~[?:?]
    at net.minecraft.world.entity.animal.Pig.travel(Pig.java:272) ~[?:?]
    at net.citizensnpcs.nms.v1_18_R2.entity.PigController$EntityPigNPC.h(PigController.java:218) ~[Citizens-2.0.30-b2565.jar:?]
    at net.minecraft.world.entity.LivingEntity.aiStep(LivingEntity.java:3307) ~[?:?]
    at net.minecraft.world.entity.Mob.aiStep(Mob.java:628) ~[?:?]
    at net.minecraft.world.entity.AgeableMob.aiStep(AgeableMob.java:154) ~[?:?]
    at net.minecraft.world.entity.animal.Animal.aiStep(Animal.java:62) ~[?:?]
    at net.minecraft.world.entity.LivingEntity.tick(LivingEntity.java:2917) ~[?:?]
    at net.minecraft.world.entity.Mob.tick(Mob.java:396) ~[?:?]
    at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1134) ~[?:?]
    at net.minecraft.server.level.WorldServer$$Lambda$8946/0x000000080285e9a0.accept(Unknown Source) ~[?:?]
    at net.minecraft.world.level.Level.guardEntityTick(Level.java:936) ~[?:?]
    at net.minecraft.server.level.ServerLevel.lambda$tick$5(ServerLevel.java:673) ~[?:?]
    at net.minecraft.server.level.WorldServer$$Lambda$8819/0x0000000802847aa0.accept(Unknown Source) ~[?:?]
    at net.minecraft.world.level.entity.EntityTickList.forEach(EntityTickList.java:43) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:653) ~[?:?]
    at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1595) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:483) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1458) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1227) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317) ~[paper-1.18.2.jar:git-Paper-"1790528"]
    at net.minecraft.server.MinecraftServer$$Lambda$4073/0x000000080143f9d0.run(Unknown Source) ~[?:?]
    at java.lang.Thread.run(Thread.java:831) ~[?:?]
```

There's also an IllegalPluginAccessException on shutdown, but that appears to be unrelated.

**Comment by mcmonkey4eva** [comment]
This has been reported on Discord a few times but nobody that's reported it has thus far been willing to post a log with debug on for some reason.
Please enable `debug-mode: true` in the `Citizens/config.yml` and post a full log replicating the issue.

If you're able to spend some time testing, it'd be very helpful to see if resetting anything stops the issue. Notably resetting your Citizens config to default (if it's been modified), or your Paper/Spigot/Bukkit configs, or your server.properties, or your world settings for the affected world. (As in, find the thing that when reset makes the issue go away, so we can narrow down that the issue only happens with whichever specific settings enabled).

-----

The logic of the issue known thus far is:
- The out-of-memory is just one of many possible ways it can manifest, but it originates from the logic for entity movement loading too much block data
- The reason it's loading too much is because it's trying to trace movement at massive speeds, so it's running a collision check across billions of blocks all at once
- It seems to consistently involve pigs? Not sure if that's coincidence or not, I think there were other mob types involved in one of the prior reports?
- It only applies to certain worlds, usually ones loaded via Multiverse though I believe there's been replication with other world-loader plugins
- It's usually been a mob NPC with waypoints directly at the spawn area of the world
- It might somehow relate to chunk loading in the world

**Comment by mcmonkey4eva** [comment]
Also for the sake of having information in one place,
Here's the first report log from PopFizz_: https://paste.denizenscript.com/View/96325
Here's the additional report log from wolf0023: https://paste.denizenscript.com/View/97211
Here's replication log from Icecapade: https://paste.denizenscript.com/View/96386 (see also Discord conversation at https://discord.com/channels/315163488085475337/971163586909073469/972224992555184229 )

**Comment by PikaMug** [comment]
With `debug-mode:true`:
https://paste.denizenscript.com/View/97405

Haven't reset anything, but I've discovered that toggling `keepSpawnInMemory` to `false` for the pig's world in _/plugins/Multiverse-Core/worlds.yml_ avoids—or at least reduces the prevalence of—the error. There is also a Horse NPC within the same area of spawn which has not produced an error, so it does seem to be just the Pig affected.

**Comment by mcmonkey4eva** [comment]
Relevant parts at line 529 and line 600:
```
[Citizens] Couldn't spawn 8{Percy, PIG} entity not added to world
[Citizens] Couldn't spawn 0{Quest Master, PLAYER} entity not added to world
[ERROR]: Entity threw exception at hub:1110.5219892492594,78.0,-719.6656556592735
(server shuts down)
[Citizens] Tried to despawn 8{Percy, PIG} while already despawned, DespawnReason.RELOAD
[Citizens] Despawned 0{Quest Master, null} DespawnReason.RELOAD
[WARN]: org.bukkit.plugin.IllegalPluginAccessException:
Plugin attempted to register Spawned{true} while not enabled
```

It's definitely from a pig (per the stacktrace), and there's no other pig NPC, so it'd have to be "Percy" (or potentially a vanilla mob and the NPC is unrelated?)

That "attempted to register Spawned" message is very suspicious as well - *might* be a result of the spawning failure, might be something else? Not sure.

cc @fullwall

**Comment by fullwall** [comment]
If possible could you try replicate with the new build?+

**Comment by PikaMug** [comment]
Using build 2568 with Multiverse's `keepSpawnInMemory: 'true'`, console reported:
`[Server] INFO [Citizens] Couldn't spawn 8{Percy, PIG} entity not added to world`

As expected, Percy did not appear in-game. On next startup with `keepSpawnInMemory` back off, Percy was restored but a duplicate of Quest Master (Player NPC) with the name "CIT-8668d2a43279" was present. Running `/citizens reload` took care of that, but now subsequent restarts no longer spawn Percy and the duplicate Quest Master returns. Again `/citizens reload` fixes both but it appears only temporary until the next restart.

**Comment by PikaMug** [comment]
Despite no plugins added during testing, aforementioned duplicate seems to be in relation to:
https://wiki.citizensnpcs.co/Frequently_Asked_Questions#My_NPCs_Have_Names_Like_CIT-abc_Instead_Of_Their_Actual_Names

Removing a color code from Quest Master's name (bringing it to 16 chars) cleared the duplicate tag. So it seems aaf870372caaef12df852e3fa6adefab71568ca7 might encourage some adverse affect on the internal scoreboard team.

**Comment by fullwall** [comment]
This sounds like a separate issue to the fast NPC? If that issue is resolved maybe split secondary issues off into a new issue?

**Comment by PikaMug** [comment]
I'm under the impression it was created as a result of the fix aaf870372caaef12df852e3fa6adefab71568ca7 (as it was never an issue prior) but I'm fine with limiting NPC names if we decide to update to a newer build. Can't speak for other users. Thanks for your help!
